File: server/channels/api4/channel_relationships.go
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package api4

import (
	"encoding/json"
	"net/http"

	"github.com/mattermost/mattermost/server/public/model"
	"github.com/mattermost/mattermost/server/public/shared/mlog"
)

func (api *API) InitChannelRelationships() {
	api.BaseRoutes.Channel.Handle("/relationships", api.APISessionRequired(getChannelRelationships)).Methods(http.MethodGet)
}

func getChannelRelationships(c *Context, w http.ResponseWriter, r *http.Request) {
	c.RequireChannelId()
	if c.Err != nil {
		return
	}

	// Get the channel to verify it exists and check permissions
	channel, err := c.App.GetChannel(c.AppContext, c.Params.ChannelId)
	if err != nil {
		c.Err = err
		return
	}

	// Check that user has permission to read the channel
	if !c.App.SessionHasPermissionToReadChannel(c.AppContext, *c.AppContext.Session(), channel) {
		c.SetPermissionError(model.PermissionReadChannelContent)
		return
	}

	// Get related channels
	response, appErr := c.App.GetRelatedChannels(c.AppContext, c.Params.ChannelId, c.AppContext.Session().UserId)
	if appErr != nil {
		c.Err = appErr
		return
	}

	if err := json.NewEncoder(w).Encode(response); err != nil {
		c.Logger.Warn("Error while writing response", mlog.Err(err))
	}
}

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/channel_relationships.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ChannelRelationshipsState} from '@mattermost/types/channel_relationships';
import type {GlobalState} from '@mattermost/types/store';

const EMPTY_RELATIONSHIPS = {};

export const getChannelRelationships = (state: GlobalState, channelId: string): ChannelRelationshipsState['byChannelId'][string] => {
    const relationships = state.entities.channelRelationships.byChannelId[channelId];

    if (!relationships) {
        return EMPTY_RELATIONSHIPS;
    }

    return relationships;
};

export const getChannelRelationship = (state: GlobalState, channelId: string, relationshipId: string) => {
    return getChannelRelationships(state, channelId)[relationshipId];
};
